### PR TITLE
Fix reshape tool missing topological points

### LINKED ISSUE #49312: Topological editing with avoid overlap creates vertices in different ways for new and edited features
### What is the bug or the crash?

When topological editing and avoid overlap is enabled:

- drawing a new overlapping polygon on first layer creates vertices automatically on the second layer (overlap is clipped, second layer feature has new vertices created along the segment where features meet)
- editing a feature on the first layer by dragging an existing vertex over a feature on the second layer does not create vertices on the second layer (overlap is clipped, second layer feature has no vertex along the segment where features meet)
- editing a feature on the first layer with reshape tool to overlap a feature on the second layer does not create vertices on the second layer (overlap is clipped, second layer feature has no vertex along the segment where features meet)

### Steps to reproduce the issue

- Open [testing.zip](https://github.com/qgis/QGIS/files/9083119/testing.zip) project file (open from - geopackage)
- Enabled editing on both vector layers
- Digitize a new feature on first layer that overlaps the second layer feature, observe that overlap is clipped and topological vertices are created
  - image
- Drag the midpoint of the segment on the northernmost feature with the vertex tool to overlap the second layer feature, observe that overlap is clipped but topological vertices are not created
  - image
- Reshape the southernmost feature to overlap the second layer feature, observe that overlap is clipped but topological vertices are not created
  - image

### Versions

QGIS version
3.22.8-Białowieża
QGIS code revision
8d5e9761df
Qt version
5.15.3
Python version
3.9.5
Compiled against GDAL/OGR
3.5.0
Running against GDAL/OGR
3.5.1
PROJ version
9.0.1
EPSG Registry database version
v10.064 (2022-05-19)
GEOS version
3.10.3-CAPI-1.16.1
SQLite version
3.38.1
PDAL version
2.3.0
PostgreSQL client version
14.3
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

Active Python plugins

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

It is possible to use reshape tool with snapping on segment enabled, create a vertex on the reshape rubberband on the segment, and a topological vertex will be created when reshape is done, regardless whether geometry precision is enabled or not. For vertex tool edits this only works if no precision is set on the edited layer, see separate issue https://github.com/qgis/QGIS/issues/49311

### PR DESCRIPTION
**Pull request by troopa81**
Add a class to avoid duplicated code when using avoid intersection and topological editing.

Use that class for vertex tool and reshape tool. It embeds modification from https://github.com/qgis/QGIS/pull/54467 .

As a consequence, last part of issue https://github.com/qgis/QGIS/issues/49312 is now fixed.

Fixes https://github.com/qgis/QGIS/issues/49312

**Funded by GISPO**

### CODE REVIEW

**Comment by troopa81** [comment]
@m-kuhn @uclaros May I request your review here ?

File: src/app/qgsmaptoolreshape.cpp [hunk 1]
```diff
@@ -182,29 +185,13 @@ void QgsMapToolReshape::reshape( QgsVectorLayer *vlayer )
...
-          {
-            res = geom.avoidIntersectionsV2( QgsProject::instance()->avoidIntersectionsLayers(), ignoreFeatures );
-            if ( res == Qgis::GeometryOperationResult::InvalidInputGeometryType )
-            {
-              emit messageEmitted( tr( "An error was reported during intersection removal" ), Qgis::MessageLevel::Critical );
-              vlayer->destroyEditCommand();
-              stopCapturing();
-              return;
-            }
+            emit messageEmitted( tr( "An error was reported during intersection removal" ), Qgis::MessageLevel::Critical );
```
Comment (m-kuhn) [inline src/app/qgsmaptoolreshape.cpp:191]: Can we make this a warning? I think yellow is severe enough for this.

**Review by m-kuhn — COMMENTED** [review]
Thank you @troopa81 it looks good, only some minor comments left.

File: src/app/maptools/qgsavoidintersectionsoperation.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/***************************************************************************
+    qgsavoidintersectionsoperation.h
+    ---------------------
+    begin                : 2023/09/20
```
Comment (lbartoletti) [inline src/app/maptools/qgsavoidintersectionsoperation.h:4]:
```suggestion
    begin                : 2023-09-20
```

File: src/app/maptools/qgsavoidintersectionsoperation.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        : operationResult( Qgis::GeometryOperationResult::NothingHappened ),
+          geometryHasChanged( false ) {}
+
+      /**
+       * avoid intersection operation result
+       */
+      Qgis::GeometryOperationResult operationResult;
+
+      /**
+       * TRUE if the geometry has changed during the avoid intersection operation
```
Comment (lbartoletti) [inline src/app/maptools/qgsavoidintersectionsoperation.h:58]:
```suggestion
       * True if the geometry has changed during the avoid intersection operation
```

File: src/app/maptools/qgsavoidintersectionsoperation.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+       */
+      bool geometryHasChanged;
+    };
+
+    /**
+     * Apply the avoid intersection operation to the geometry \a geom associated to the feature \a fid
+     * from the layer \a layer.
+     * \param ignoreFeatures possibility to give a list of features where intersections should be ignored (not available in Python bindings)
```
Comment (lbartoletti) [inline src/app/maptools/qgsavoidintersectionsoperation.h:66]: Is only thsie parameter not available in Python or the whole method?

**Review by lbartoletti — COMMENTED** [review]
lgtm with minor style changes

File: src/app/CMakeLists.txt [hunk 5]
```diff
@@ -229,6 +229,7 @@ set(QGIS_APP_SRCS
   locator/qgslocatoroptionswidget.cpp
 
   maptools/qgsappmaptools.cpp
+  maptools/qgsavoidintersectionsoperation.cpp
```
[the remark this answers is not in the document]
Comment (troopa81) [reply src/app/CMakeLists.txt:232]: No headers are explicitly described in app module as there is no need to install them.

File: src/app/maptools/qgsavoidintersectionsoperation.h (hunk 4 above)
Comment (troopa81) [reply src/app/maptools/qgsavoidintersectionsoperation.h:66]: There is no python code as this class is in app, I removed the comments

File: src/app/qgsmaptoolreshape.cpp (hunk 1 above)
Comment (troopa81) [reply src/app/qgsmaptoolreshape.cpp:191]: Yes. Done

File: src/app/maptools/qgsavoidintersectionsoperation.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        } );
+        messageBarItem->layout()->addWidget( restoreButton );
+        QgisApp::instance()->messageBar()->pushWidget( messageBarItem, Qgis::MessageLevel::Info, 15 );
+      }
+      break;
+    }
+
+    case Qgis::GeometryOperationResult::InvalidBaseGeometry:
+      result.geometryHasChanged = true;
+      emit messageEmitted( tr( "At least one geometry intersected is invalid. These geometries must be manually repaired." ), Qgis::MessageLevel::Warning );
```
[the remark this answers is not in the document]
Comment (troopa81) [reply src/app/maptools/qgsavoidintersectionsoperation.cpp:101]: >  (I think the current wording could be improved)

"While avoiding intersection, at least one of the geometries involved is invalid." ?

**Review by m-kuhn — APPROVED** [review]

**Review by lbartoletti — APPROVED** [review]

**Comment by troopa81** [comment]
Thanks for reviewing @m-kuhn and @lbartoletti

**Pull request merged**